File: python/src/main/python/Q054.py
from src.main.python.Solution import Solution

# Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.
#
# For example,
#   Given the following matrix:
#
#   [
#     [ 1, 2, 3 ],
#     [ 4, 5, 6 ],
#     [ 7, 8, 9 ]
#   ]
#   You should return [1,2,3,6,9,8,7,4,5].


class Q054(Solution):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        def spiral_order(matrix, row, col, offset, ans):
            if row <= 0 or col <= 0:
                return
            elif row == 1:
                for j in range(col):
                    ans.append(matrix[offset][offset+j])
            elif col == 1:
                for i in range(row):
                    ans.append(matrix[offset+i][offset])
            else:
                for j in range(col-1):
                    ans.append(matrix[offset][offset+j])
                for i in range(row-1):
                    ans.append(matrix[offset+i][offset+col-1])
                for j in range(col-1):
                    ans.append(matrix[offset+row-1][offset+col-1-j])
                for i in range(row-1):
                    ans.append(matrix[offset+row-1-i][offset])
                spiral_order(matrix, row-2, col-2, offset+1, ans)

        ans = []
        if matrix and matrix[0]:
            spiral_order(matrix, len(matrix), len(matrix[0]), 0, ans)
        return ans

File: java/src/main/java/solutions/Q054.java
package solutions;

import java.util.ArrayList;
import java.util.List;

/**
 * Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.
 *
 * For example,
 *  Given the following matrix:
 *
 *  [
 *      [ 1, 2, 3 ],
 *      [ 4, 5, 6 ],
 *      [ 7, 8, 9 ]
 *  ]
 *
 *  You should return [1,2,3,6,9,8,7,4,5].
 */

public class Q054 extends Solution {

    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        if (matrix != null && matrix.length != 0 && matrix[0].length != 0) {
            spiralOrder(matrix, matrix.length, matrix[0].length, 0, ans);
        }
        return ans;
    }

    private void spiralOrder(int[][] matrix, int m, int n, int k, List<Integer> ans) {
        if (m > 0 && n > 0) {
            if (m == 1) {
                for (int j = 0; j < n; ++j) {
                    ans.add(matrix[k][j+k]);
                }
            } else if (n == 1) {
                for (int i = 0; i < m; ++i) {
                    ans.add(matrix[k+i][k]);
                }
            } else {
                for (int j = 0; j < n - 1; ++j) {
                    ans.add(matrix[k][k+j]);
                }
                for (int i = 0; i < m - 1; ++i) {
                    ans.add(matrix[k+i][k+n-1]);
                }
                for (int j = 0; j < n - 1; ++j) {
                    ans.add(matrix[k+m-1][k+n-1-j]);
                }
                for (int i = 0; i < m - 1; ++i) {
                    ans.add(matrix[k+m-1-i][k]);
                }
                spiralOrder(matrix, m-2, n-2, k+1, ans);
            }
        }
    }

}
